### ISSUE TITLE
Layout shift when routing to different page

### ISSUE THREAD

**Issue by AlexLup06** [vercel/next.js#43418]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

    Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 21.6.0: Mon Aug 22 20:19:52 PDT 2022; root:xnu-8020.140.49~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.14.2
      npm: 8.5.0
      Yarn: 1.22.18
      pnpm: N/A
    Relevant packages:
      next: 13.0.6-canary.1
      eslint-config-next: 13.0.5
      react: 18.2.0
      react-dom: 18.2.0

### What browser are you using? (if relevant)

Chrome 107.0.5304.87

### How are you deploying your application? (if relevant)

next dev

### Describe the Bug

When I click on Link to get to a new page, there is this very short but visible layout shift. This only happens when using CSS Modules. But as the saying goes:  A Picture  is worth more than a thousand words.

**With CSS modules:**

![Bildschirmfoto 2022-11-26 um 22 54 40](https://user-images.githubusercontent.com/87496469/204110430-aa2b9da8-cd9f-48bc-ba09-4d529667614d.png)

Shows loading page without styles:
![Bildschirmfoto 2022-11-26 um 22 54 48](https://user-images.githubusercontent.com/87496469/204110435-374ea858-6d9c-479a-9d5b-afedfa6c32ac.png)

Suddenly loads styles:
![Bildschirmfoto 2022-11-26 um 22 54 57](https://user-images.githubusercontent.com/87496469/204110437-8491aa2f-6b40-40e5-ba22-b6ba62ecf557.png)

Now page content arrives but loads without styles:
![Bildschirmfoto 2022-11-26 um 22 55 07](https://user-images.githubusercontent.com/87496469/204110443-3c75fe2b-21cf-4007-af72-1b68e6bd93da.png)

Suddenly loads styles:
![Bildschirmfoto 2022-11-26 um 22 55 20](https://user-images.githubusercontent.com/87496469/204110445-bfe70c6b-c77f-4b8b-ab2e-28eaddf32d7c.png)

---------------------------------------

**With inline styling:**

![Bildschirmfoto 2022-11-26 um 23 00 59](https://user-images.githubusercontent.com/87496469/204110507-1015a34f-ccbe-419d-8959-e954cdf039eb.png)

Shows loading page with styles right from the start:
![Bildschirmfoto 2022-11-26 um 23 01 07](https://user-images.githubusercontent.com/87496469/204110528-e1938ca9-e0a2-42a6-b0f8-a80e67728b14.png)

Shows page content with styles right from the start:
![Bildschirmfoto 2022-11-26 um 23 01 14](https://user-images.githubusercontent.com/87496469/204110545-5d0d08b7-af0c-4ef9-8bc6-b4b8b76253b2.png)

And trust me the layout shift looks even more trashy IRL. But as you can see the shift does not happen with inline styling but with css modules. 
Overall css modules kinda suck in nextjs 13. There are also multiple issues open with css modules not working in pages directory when using` appDir:true` in next.config.js .... experienced it myself.

My guess is the above bug I presented is not a feature but actually a bug?

### Expected Behavior

Don't do any layout shifts and apply styles right from the start when using css modules.

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://codesandbox.io/s/adoring-bogdan-inutd7?file=/app/page.tsx

### To Reproduce

Click on the Link to get to example. /test1 shows inline styling and /test2 shows css modules.

**Comment by Fredkiss3** [comment]
I've stumbled upon a similar problem but not with css modules, but just regular css import, where i tried to import a css from a child component and i kept having unexplicable bugs : the page would be stuck at the loading screen.

My solution was to remove the import from the child component and import the css directly from the root layout. I guess it's a problem with React Suspense and streaming, where it needs to transform the css import into a link in the `<head>` tag, it messes up the hydration or something...

It's not exactly the same problem, but i think they have the same root, when you use a `loading.tsx` file,React renders first your loading screen, then the component beneath, so the css module compilation is on the fly, and when the compilation has finished, you have your styles applied.

I don't know if a workaround for css modules is even possible.

**Comment by rafma0** [comment]
Yes, coming from styled-components to css modules I also noticed some "weird" behaviors of page elements taking a while to show the correct style. Even if it takes 1 sec or less to load the .css, it gets noticeable in some cases.

The thing is, the .css is only added to the `<head>` the first time the user requests something importing it. Its really fast but it can look weird for some elements.

My solution was similar to what @Fredkiss3 said. You need your **critical** styles loaded before the new page needs it, so you can put them in your layout.module.scss, `import 'layout.module.scss'` in your `layout.tsx` and then in your page you do something like `import layoutStyles from '../layout.module.scss'` and it will use the .css loaded before. Or you can create a `shared.module.scss`, etc.

I'm also not sure there's a way around it, unless you can some way make pages with inlined styles from .css modules.

**Comment by Fredkiss3** [comment]
The thing with css modules is that you want to use them where you import them and i suppose it is scoped by the file which imports them ? 🤔

If you import your styles in the root, at this point you why not to use regular css or scss ?

**Comment by rafma0** [comment]
> The thing with css modules is that you want to use them where you import them and i suppose it is scoped by the file which imports them ? 🤔
> 
> If you import your styles in the root, at this point you why not to use regular css or scss ?

From what I could understand, when you import a css module what you get is the unique class names that were created when the .css was compiled. Also, the compiled .css gets added to your `<head>` the first time a layout / page / component importing it is requested.

If you were to import your module in the root layout and in every other page, I guess it wouldn't make sense to use modules, like you said. But you can import a module in specific sub layouts / pages and share those class names only with other files you choose.

In my case I have a `.title` in a module, and I imported the module in the sub layout and in the pages extending this layout, so I don't get layout shifts when the user navigate to the pages for the first time. But I also have a `page.module.scss` for each page, for the rest of the styles.

Idk if what I'm doing is "correct" but I honestly I can't think of any other way right now. It avoids redefining the same class in multiple .css and also solves the layout shift since the .css is loaded by the layout before the page is shown.

**Comment by Fredkiss3** [comment]
what about sub routes with a `loading.tsx` which have also their own layout ? 
does it still work without a layout shift ?

i'm talking about routes like this : 
```
route/
   loading.tsx
   sub-route/
        layout.tsx
        page.tsx
        page.module.css
        layout.module.css
```

**Comment by rafma0** [comment]
> what about sub routes with a `loading.tsx` which have also their own layout ? does it still work without a layout shift ?

In my tests, having a `loading.tsx` or not, does not make a difference for when the new .css is added to the `<head>` (I couldn't replicate what @AlexLup06 showed, where the loading message shows not styled, then styled, then the page does the same)

in your example, if you load first `/route` then navigate to `/route/sub-route`, for me it will replace the contents of the `/route` page with the `loading.tsx`, then show the new page and after all that, add the 2 new .css to your `<head>`, so depending on your layout you can see a shift.

I guess if you have complex pages with multiple sub routes / layouts, doing weird shifts, you have to come up with strategic places to load some of your styles. Like for the `/route/sub-route` maybe the critical styles for `sub-route` could be put in the `/route/layout.module.css` ?

You could also mix regular .css and modules, and put only layout breaking stuff as a regular .css loaded in your root, maybe in your `globals.css`.

There's also a big chance that I'm overcomplicating everything 😄

**Comment by AlexLup06** [comment]
Hey guys,
Thanks for discussing this issue!
It seems like something does not work with the reproduction link I provided.
Here is a repo: https://github.com/AlexLup06/layoutshift_bug_next13.git

The issue still exists... In chrome dev tools you can throttle your network connection to slow 3G to see the effects in real time.

The issue also exists in production but it only happens there sometimes because of the caching mechanism. If you restart Chrome you can see the effect when you go to `/inlinestyles` and click the link (again with slow 3G or recording the screen in performance tab to later see the effect).

**Comment by rafma0** [comment]
Oh, I see now, your loading is also importing a module so it does the same thing as the page.

For the loading part its hard to see even with Slow 3G enabled, but it would botter me too. You can make it easier to see it, changing your page func to async and adding `await new Promise(resolve => setTimeout(resolve, 5000))` before the return.

I recorded a quick video because why not, I also want to learn the best way to solve this :)


https://user-images.githubusercontent.com/28578847/205440000-290c14b8-e3eb-4811-a613-8795e912d48c.mp4

**Comment by Fredkiss3** [comment]
It seems to me that you would have to not use css modules since they are evaluated at the time they are imported. See, when your page is loading, it is also evaluating the css to import and load them after.

The best way In my Opinion to mitigate this is to use your modules in your layout files, since normally your layout files are not evaluated after…

also what is the result in production mode ? Does it still produce a layout shift ?

**Comment by AlexLup06** [comment]
@Fredkiss3 

> The issue also exists in production but it only happens there sometimes because of the caching mechanism. If you restart Chrome you can see the effect when you go to `/inlinestyles` and click the link (again with slow 3G or recording the screen in performance tab to later see the effect).

**Comment by Fredkiss3** [comment]
Can you generate the css for the modules ahead of time ?
Like running a different process that analyse all your component’s files and generate all the css ? 🤔
